python/GangaDirac/Lib/Backends/DiracBase.py

```diff
@@ -974,8 +957,26 @@ def updateMonitoringInformation(jobs_):
         #logger.debug('Monitor jobs    : ' + repr([j.fqid for j in monitor_jobs]))
         #logger.debug('Requeue jobs    : ' + repr([j.fqid for j in requeue_jobs]))
 
+        # Split all the monitorable jobs into groups based on the
+        # credential used to communicate with DIRAC
+        # TODO Perhaps this would make more sense being done on a more generic level
+        monitor_jobs_by_credential = defaultdict(list)  # type: Dict[ICredentialInfo, List[Job]]
+        for j in monitor_jobs:
+            try:
+                cred_req = j.backend.credential_requirements
+                cred = credential_store[cred_req]
+                if not cred.is_valid():
+                    logger.debug('Required credential %s is not valid', cred)
+                    needed_credentials.add(cred_req)
+                    continue
+                monitor_jobs_by_credential[cred].append(j)
+            except KeyError:
```

Comment:
I'm not sure it will be any faster as the standard definition of `__contains__` is:
```python
def __contains__(self, key):
    try:
        self[key]
    except KeyError:
        return False
    else:
        return True
```
The point is that in the `try` block you should be able to write code as if there was no problem without checking and then use the `except` block for situations where that assumption failed (see 'EAFP' in https://docs.python.org/2/glossary.html).

The throwing and catching of exceptions as logic flow is generally considered idiomatic in Python, as long as it's not overdone.

That said, I will look at rewriting this part in that style and see if it's more readable.